module/zfs/zio.c

```diff
@@ -4424,6 +4492,100 @@ zio_alloc_zil(spa_t *spa, objset_t *os, uint64_t txg, blkptr_t *new_bp,
 	return (error);
 }
 
+/*
+ * ==========================================================================
+ * Read and write to the logical top device
+ * ==========================================================================
+ */
+/*
+ * Logical reads and writes arrive at zio_logical_io_start(), which issues the
+ * first child IO to specific vdevs (through zio_vdev_child_io()). The logical
+ * IO will wait at zio_logical_io_done() for the tree of child IO to complete,
+ * and then process its result.
+ *
+ * Thus, these represent the handoff between the logical IO pipeline
+ * (compression, encryption, checksums, DVA allocation, etc) and the physical
+ * IO pipeline (the vdev tree).
+ */
+
+static void
+zio_logical_io_child_done(zio_t *zio)
+{
+	zio_t *pio = zio->io_private;
+	pio->io_error = zio->io_error;
```

Comment:
May be out of this PR scope, but while I see similar code in random vdev types implementation, it seems to duplicate `zio_inherit_child_errors()` calls in `zio_done()` and a code in `zio_notify_parent()`.

Reply:
Yeah. It could be achieved by not setting `ZIO_FLAG_DONT_PROPAGATE`, but I haven't looked hard into the side-effects of that.

If nothing else, use of an `io_done` callback "inside" the pipeline smells a bit off to me.

I do have in my notes "why don't vdev child IOs propagate errors by default?". I think I would like to revisit that sometime.

But yeah, maybe another time.